runtime/src/consensus/tendermint/verifier.rs

```diff
@@ -765,14 +809,110 @@ impl Verifier {
                 }
             }
 
-            // Persist trusted root every once in a while.
-            if cache.last_trust_root.height - last_saved_trust_root_height
-                > TRUST_ROOT_SAVE_INTERVAL
-            {
-                self.save_trust_root(&untrusted_local_store, &cache);
-                last_saved_trust_root_height = cache.last_trust_root.height;
+            // Persist last verified block once in a while.
+            let last_height = cache.latest_known_height();
+            if last_height - last_saved_verified_block_height > TRUSTED_STATE_SAVE_INTERVAL {
+                self.save_trusted_state(&cache.last_verified_block, &untrusted_local_store);
+                last_saved_verified_block_height = last_height;
+            }
+        }
+    }
+
+    fn handle_chain_context_change(
+        &self,
+        trusted_state: TrustedState,
+        verifier: &impl TMVerifier,
+        clock: &impl components::clock::Clock,
+        io: &impl components::io::Io,
+    ) -> Result<TrustRoot, Error> {
+        let host_info = self.protocol.get_host_info();
+
+        // Nothing to handle.
+        if trusted_state.trust_root.chain_context == host_info.consensus_chain_context {
+            info!(self.logger, "Consensus chain context hasn't changed");
+            return Ok(trusted_state.trust_root);
+        }
+        info!(self.logger, "Consensus chain context has changed");
+
+        // Chain context transition cannot be done directly from the embedded
+        // trust root as we don't have access to the matching trusted light
+        // block which validator set we need to verify blocks from the new chain.
+        let trusted_block = trusted_state
+            .trusted_block
+            .ok_or(Error::ChainContextTransitionFailed)?;
+
+        // Fetch the latest block and prepare untrusted state for verification.
+        let untrusted_block = io
+            .fetch_light_block(components::io::AtHeight::Highest)
```

Comment:
It would be great if this could be done at genesis as the validator set could change between genesis and highest. Note that in case someone forgets to upgrade a node for a while the difference between genesis and highest could be potentially large. The question is how to determine genesis of the new network.

Reply:
Changed. I used the following approach. First I ask the host for genesis height, then I fetch the genesis block and finally I verify if this block is the first one (`last_block_id` in the `header` should be `None`).